packages/nightingale-track-canvas/src/utils/utils.ts

```diff
@@ -0,0 +1,208 @@
+            running = false;
+        }
+    }
+    return {
+        requestRefresh,
+    };
+}
+
+
+/** Sleep for `ms` milliseconds. */
+export function sleep(ms: number): Promise<void> {
+    return new Promise<void>(resolve => setTimeout(() => {
+        resolve();
+    }, ms));
+}
+
+
+/** Return the last element of `array`, or `undefined` if there are no elements.
+ * If `predicate` is provided, return the last element where `predicate` returns true,
+ * or `undefined` if there is no such element. */
+export function last<T>(array: T[], predicate?: (value: T, index: number, obj: T[]) => boolean): T | undefined {
+    if (!predicate) return array.length > 0 ? array[array.length - 1] : undefined;
+    for (let i = array.length - 1; i >= 0; i--) {
+        const element = array[i];
+        if (predicate(element, i, array)) return element;
+    }
+    return undefined;
+}
+
+
+/** Return index of the first element of `sortedArray` for which `key(element) >= query`.
+ * Return length of `sortedArray` if `key(element) < query` for all elements.
+ * (aka Return the first index where `query` could be inserted while keeping the array sorted.) */
+export function firstGteqIndex<T>(sortedArray: ArrayLike<T>, query: number, key: (element: T) => number) {
+    return firstGteqIndexInRange(sortedArray, query, 0, sortedArray.length, key);
+}
+
+/** Return index of the first element within range [start, end) for which `key(element) >= query`.
+ * Return `end` if for all elements in the range `key(element) < query`. */
+function firstGteqIndexInRange<T>(sortedArray: ArrayLike<T>, query: number, start: number, end: number, key: (element: T) => number) {
+    if (start === end) return start;
+    // Invariants:
+    // key(sortedArray[i]) < query for each i < start
+    // key(sortedArray[i]) >= query for each i >= end
+    while (end - start > 4) {
+        const mid = (start + end) >> 1;
+        if (key(sortedArray[mid]) >= query) {
+            end = mid;
+        } else {
+            start = mid + 1;
+        }
+    }
+    // Linear search remaining 4 or fewer elements:
+    for (let i = start; i < end; i++) {
+        if (key(sortedArray[i]) >= query) {
+            return i;
+        }
+    }
+    return end;
+}
+
+
+/** Data structure for storing integer intervals (ranges) and efficiently retrieving a subset of ranges which overlap with another (query) interval.
+ * Not suited for storing float intervals, but query interval can be float. */
+export class RangeCollection<T> {
+    protected readonly items: T[];
+    protected readonly starts: number[];
+    protected readonly stops: number[];
+
+    /** Keys to `this.bins`, sorted in ascending order */
+    protected readonly binSpans: number[];
+    /** `this.bins[span]` contains indices of all items whose length is `<= span` but `> span/Q`, in ascending order */
+    protected readonly bins: Record<number, number[]>;
+    /** Ratio of spans of neighboring bins */
+    protected readonly Q = 2;
```

Comment:
Curious how 2 was selected?

Reply:
Could be any number >1 in theory. But 2 works nice and gives easy-to-work-with spans (1,2,4,8,16...).
With higher ratios you have weak assumption about item length, so you need to check more items that are in fact not overlapping -> worse performance.
On the other hand, smaller ratios (e.g. 1.5) could produce too many bins, which would be more likely to be empty -> worse performance; plus the spans would be really ugly (1,1.5,2.25, 3.375...).

To be more rigorous, the optimal value Q depends on the expected number of items and sequence length (which we don't know exactly so I picked a nice number :D)